### PR TITLE
Add L_TRANSFER_ADJUSTMENTS table

### PR DESCRIPTION
**Pull request by clark800**

### CODE REVIEW

File: src/lib/db.js [hunk 1]
```diff
@@ -6,12 +6,14 @@ const path = require('path')
...
 const spawn = require('child_process').spawn
 const knex = require('./knex').knex
+const sequence = require('./utils').sequence
 const readRejectionReasons = require('../models/db/rejectionReasons')
   .readRejectionReasons
 const readTransferStatuses = require('../models/db/transferStatuses')
   .readTransferStatuses
 
 const TABLE_NAMES = [
+  'L_TRANSFER_ADJUSTMENTS',
```
Comment (MatthewPhinney) [inline src/lib/db.js]: Do you need to include the lookup tables in this list?

File: src/models/db/accounts.js [hunk 2]
```diff
@@ -66,6 +67,18 @@ function getAccountByFingerprint (fingerprint, options) {
   return db.selectOne({FINGERPRINT: fingerprint}, options && options.transaction)
 }
 
+function getAccountById (id, options) {
+  return db.selectOne({ACCOUNT_ID: id}, options && options.transaction)
+}
+
+function getAccountId (name, options) {
+  const transaction = options && options.transaction
+  return (transaction || knex).from(TABLE_NAME).select()
```
Comment (MatthewPhinney) [inline src/models/db/accounts.js]: Can you not use `getAccount` here?

File: src/lib/db.js (hunk 1 above)
Comment (clark800) [inline src/lib/db.js]: No, but I think it's ok the leave it this way since it's simpler to just have a list of all table names.

File: src/models/db/accounts.js (hunk 2 above)
Comment (clark800) [inline src/models/db/accounts.js]: Yeah because getAccount doesn't return the id

Comment (MatthewPhinney) [inline src/models/db/accounts.js]: Oh, I see.

File: src/models/db/adjustments.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+const db = require('./utils')()
+
+function convertFromPersistentAdjustment (data, options) {
+  return getAccountById(data.ACCOUNT_ID, options).then((account) => {
+    return _.omit({
+      account: account.name,
+      amount: Number(data.AMOUNT).toString(),
+      authorized: Boolean(data.IS_AUTHORIZED),
+      memo: data.MEMO
+    }, (x) => !x)
```
Comment (MatthewPhinney) [inline src/models/db/adjustments.js]: This omits falsy properties?

File: src/models/db/adjustments.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  const debits = Promise.all(rows.filter((row) => row.DEBIT_CREDIT === 'debit')
+    .map((debit) => convertFromPersistentAdjustment(debit, options)))
+  const credits = Promise.all(rows.filter((row) => row.DEBIT_CREDIT === 'credit')
+    .map((credit) => convertFromPersistentAdjustment(credit, options)))
+  return Promise.all([debits, credits]).then((results) => {
+    if (results[0].length === 0 && results[1].length === 0) {
+      return {}
+    }
+    return {
+      debits: _.sortBy(results[0], (adjustment) => adjustment.account),
```
Comment (MatthewPhinney) [inline src/models/db/adjustments.js]: What's the importance of the sort order here?

File: src/models/db/adjustments.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+function convertToPersistentAdjustment (transferId, type, data, options) {
+  return getAccountId(data.account, options).then((accountId) => {
+    return {
+      TRANSFER_ID: transferId,
+      ACCOUNT_ID: accountId,
+      DEBIT_CREDIT: type,
+      AMOUNT: data.amount,
+      IS_AUTHORIZED: Number(data.authorized || false),
```
Comment (MatthewPhinney) [inline src/models/db/adjustments.js]: Do we update rows in this table? If so, we need to explicitly pass `IS_AUTHORIZED` on all updates, or it will be set to `false`.

File: src/sql/pg/create.sql [hunk 6]
```diff
@@ -90,6 +88,31 @@ CREATE INDEX "L_XIF_TRANSFERS_REASON" ON "L_TRANSFERS"
...
+  "MEMO" CHARACTER VARYING(4000) NULL
+);
+
+CREATE INDEX "L_XPK_TRANSFER_ADJUSTMENTS" ON "L_TRANSFER_ADJUSTMENTS"
+  ("TRANSFER_ADJUSTMENT_ID" ASC);
+ALTER TABLE "L_TRANSFER_ADJUSTMENTS" ADD CONSTRAINT "L_PK_TRANSFER_ADJUSTMENTS"
+  PRIMARY KEY ("TRANSFER_ADJUSTMENT_ID");
+CREATE UNIQUE INDEX "L_XAK_TRANSFER_ADJUSTMENTS" ON "L_TRANSFER_ADJUSTMENTS"
+  ("TRANSFER_ID" ASC, "ACCOUNT_ID" ASC, "DEBIT_CREDIT" ASC);
+CREATE INDEX "L_XIF_TRANSFER_ADJUSTMENTS_TRA" ON "L_TRANSFER_ADJUSTMENTS"
```
Comment (MatthewPhinney) [inline src/sql/pg/create.sql]: Is the `_TRA` suffix for "transfer"?

File: src/models/db/adjustments.js (hunk 3 above)
Comment (clark800) [inline src/models/db/adjustments.js]: Oh yeah, I should change this to `_.isNil`

File: src/models/db/adjustments.js (hunk 5 above)
Comment (clark800) [inline src/models/db/adjustments.js]: Will fix

File: src/models/db/adjustments.js (hunk 4 above)
Comment (clark800) [inline src/models/db/adjustments.js]: Just needs to be consistent or else you get invalid modification errors

File: src/sql/pg/create.sql (hunk 6 above)
Comment (clark800) [inline src/sql/pg/create.sql]: Yeah

File: src/sql/strong-oracle/create.sql [hunk 7]
```diff
@@ -167,6 +173,40 @@ ALTER TABLE "L_TRANSFERS" ADD (CONSTRAINT "FK_STATUS_ID__TRANSFERS"
...
+CREATE INDEX L_XPK_TRANSFER_ADJUSTMENTS ON "L_TRANSFER_ADJUSTMENTS"
+  ("TRANSFER_ADJUSTMENT_ID" ASC);
+
+ALTER TABLE "L_TRANSFER_ADJUSTMENTS" ADD CONSTRAINT PK_TRANSFER_ADJUSTMENTS
+  PRIMARY KEY ("TRANSFER_ADJUSTMENT_ID");
+
+CREATE UNIQUE INDEX L_XAK_TRANSFER_ADJUSTMENTS ON "L_TRANSFER_ADJUSTMENTS"
+  ("TRANSFER_ID" ASC, "ACCOUNT_ID" ASC);
+
+ALTER TABLE "L_TRANSFER_ADJUSTMENTS" ADD CONSTRAINT L_XAK_TRANSFER_ADJUSTMENTS
```
Comment (MatthewPhinney) [inline src/sql/strong-oracle/create.sql]: The unique index above doesn't also add a uniqueness constraint?

File: src/sql/strong-oracle/create.sql [hunk 8]
```diff
@@ -167,6 +173,40 @@ ALTER TABLE "L_TRANSFERS" ADD (CONSTRAINT "FK_STATUS_ID__TRANSFERS"
...
+  UNIQUE ("TRANSFER_ID", "ACCOUNT_ID");
+
+CREATE INDEX L_XIF_TRANSFER_ADJUSTMENTS ON "L_TRANSFER_ADJUSTMENTS"
+  ("TRANSFER_ID" ASC);
+
+CREATE INDEX L_XIF_TRANSFER_ADJUSTMENTS ON "L_TRANSFER_ADJUSTMENTS"
+  ("ACCOUNT_ID" ASC);
```
Comment (MatthewPhinney) [inline src/sql/strong-oracle/create.sql]: What are the consequences of having both indexes (account_id and transfer_id) use same name?

File: src/sql/strong-oracle/create.sql [hunk 9]
```diff
@@ -263,6 +303,23 @@ BEGIN
...
+CREATE OR REPLACE TRIGGER L_TRG_TRANSFER_ADJUSTMENTS_SEQ
+  BEFORE INSERT
+  ON "L_TRANSFER_ADJUSTMENTS"
+  FOR EACH ROW
+  -- Optionally restrict this trigger to fire only when really needed
+  WHEN (new."TRANSFER_ADJUSTMENT_ID" is null)
+DECLARE
+  v_id "L_TRANSFER_ADJUSTMENTS"."TRANSFER_ADJUSTMENT_ID"%TYPE;
+BEGIN
+  -- Select a new value from the sequence into a local variable. As David
```
Comment (MatthewPhinney) [inline src/sql/strong-oracle/create.sql]: Who is David?

File: src/sql/strong-oracle/drop.sql [hunk 10]
```diff
@@ -1,4 +1,14 @@
 BEGIN
+   EXECUTE IMMEDIATE 'DROP TABLE "L_TRANSFER_ADJUSTMENTS"';
+EXCEPTION
+   WHEN OTHERS THEN
+      IF SQLCODE != -942 THEN
+         RAISE;
+      END IF;
+END;
+/
+
```
Comment (MatthewPhinney) [inline src/sql/strong-oracle/drop.sql]: Don't you need to also drop the indexes and triggers, or is that going to be another PR?

File: src/sql/strong-oracle/create.sql (hunk 8 above)
Comment (clark800) [inline src/sql/strong-oracle/create.sql]: They should be different, will fix

File: src/sql/strong-oracle/create.sql (hunk 7 above)
Comment (clark800) [inline src/sql/strong-oracle/create.sql]: Seems like it should, I'll check with Abraham

File: src/sql/strong-oracle/create.sql (hunk 9 above)
Comment (clark800) [inline src/sql/strong-oracle/create.sql]: This was copied from Abraham's scripts, I'll delete it

File: src/sql/strong-oracle/drop.sql (hunk 10 above)
Comment (clark800) [inline src/sql/strong-oracle/drop.sql]: I'll do that in another pull request since it should be done for all tables

File: src/sql/strong-oracle/create.sql (hunk 7 above)
Comment (clark800) [inline src/sql/strong-oracle/create.sql]: It turns out that both lines are necessary because a unique index and a unique constraint are separate concepts in Oracle: http://stackoverflow.com/questions/7521817/oracle-unique-constraint-and-unique-index-question

**Comment by clark800** [comment]
@MatthewPhinney updated

**Comment by MatthewPhinney** [comment]
LGTM
